runtime/obj.c

```diff
@@ -259,39 +260,25 @@ CAMLprim value caml_lazy_read_result (value v)
 CAMLprim value caml_lazy_update_to_forcing (value v)
 {
   tag_t tag;
-  value field0;
 
   tag = obj_tag(v);
   if (tag != Lazy_tag && tag != Forcing_tag && tag != Forward_tag) {
     /* v is not a lazy block. It must have been forced by another domain and
      * short-circuited by the GC at a safepoint. */
-    return Val_int(4);
+    return Val_int(3);
   }
 
   /* v is a lazy block with Lazy_tag, Forcing_tag or Forward_tag. */
   if (obj_update_tag (v, Lazy_tag, Forcing_tag)) {
     /* Successfully update the tag to Forcing_tag */
     return Val_int(0);
   } else {
-    field0 = Field(v,0);
-    /* The tag has to be read again after the field field. We use [atomic_load]
-     * in order to enforce ordering between the two reads. */
-    tag = Tag_hd (atomic_load (Hp_atomic_val(v)));
+    tag = obj_tag(v);
 
     if (tag == Forcing_tag) {
-      if (field0 == caml_ml_domain_unique_token(Val_unit))
         return Val_int(1);
-      else
-        return Val_int(2);
-    } else if (tag == Forward_tag) {
-      return Val_int(3);
     } else {
-      CAMLassert (tag == Lazy_tag);
-      /* We may reach here if the tag of [v] has been reset to [Lazy_tag] using
-       * [caml_lazy_reset_to_lazy] since our own attempt at updating the tag of
-       * [v] in [obj_update_tag] failed. The compare and swap in
-       * [obj_update_tag] failed since it must have observed [Forcing_tag].
-       * This represents a racy update of the tag of [v] by another mutator. */
+      CAMLassert (tag == Forward_tag);
       return Val_int(2);
     }
```

Comment:
I think you are right (the assertion `tag == Forward_tag` can fail), but this is related to this PR.

Reply:
Firstly, my overall idea with this function is to leave some of the functionality intact even though `camlinternalLazy.ml` only cares about `0` and `not 0` cases. The idea is that future implementation of concurrency safe lazy may need to distinguish the various cases. Hence, my choice was to leave the structure of the function as is. 

It is useful to distinguish `Forward` from other shortcut tags; if the result is former and you have the assurance that the GC hasn't occurred, which you currently, the code in `camlinternalLazy.ml` can read the first field of the `Forward` object and return it as a result. 

As @gadmm, the assertion is a bug. 

My assessment is that, rather than think about tricky corner cases, we may simplify the function such that it just returns `0` if the function succeeds in atomically updating the tag from `Lazy` to `Forcing`, and `1` in all the other cases. This way we get to a state where we can merge multicore MVP trunk, and further development can happen on ocaml/ocaml. Thoughts?

Reply:
Simplified `update_to_forcing` in the latest commits.

Reply:
The simplified version is very nice, thanks!

Reply:
It needs one extra check to ensure that the value we are attempting to update the tag of is actually a block. I've added this check in `update_to_forcing`.